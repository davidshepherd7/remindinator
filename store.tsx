import AsyncStorage from "@react-native-async-storage/async-storage"
import { combineReducers, configureStore, createSlice, getDefaultMiddleware } from "@reduxjs/toolkit"
import { DateTime, Duration } from "luxon"
import { FLUSH, PAUSE, PERSIST, persistReducer, PURGE, REGISTER, REHYDRATE } from "redux-persist"
import { v4 as uuid } from 'uuid';
import * as Notifications from 'expo-notifications';

function MINUTES(m: number): Duration {
    return Duration.fromObject({ minutes: m })
}

function scheduleNotify(reminder: Reminder) {
    [0, 1, 2].map(i => {
        const when = reminder.time.plus(MINUTES(10 * i))
        Notifications.scheduleNotificationAsync({
            content: {
                title: "Do the thing",
                body: 'Please :)',
                data: {
                    reminderId: reminder.id,
                    reminderNumber: i
                }
            },
            trigger: {
                hour: when.hour,
                minute: when.minute,
                repeats: true
            },
        });
    })
}


export interface Reminder {
    id: string
    time: DateTime
}

export function mkReminder(hours: string, minutes: string): Reminder {
    return {
        id: uuid(),
        time: DateTime.fromObject({
            hour: parseInt(hours), minute: parseInt(minutes)
        })
    }
}



const reminderSlice = createSlice({
    name: 'reminders',
    initialState: {
        reminders: [] as Reminder[]
    },
    reducers: {
        addReminder: (state, action) => {
            const { hours, minutes } = action.payload
            const r = mkReminder(hours, minutes)
            scheduleNotify(r)
            return { reminders: [...state.reminders, r] }
        }
    }
})
export const { addReminder } = reminderSlice.actions



const reducers = combineReducers({ reminders: reminderSlice.reducer })
const persistConfig = {
    key: 'root',
    version: 1,
    storage: AsyncStorage,
}
const persistedReducer = persistReducer(persistConfig, reducers);
export const store = configureStore({
    reducer: persistedReducer, middleware: getDefaultMiddleware({
        serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
        }
    })
})
