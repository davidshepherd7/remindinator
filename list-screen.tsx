import React from "react";
import { View, Text } from "react-native";
import { useSelector } from "react-redux";
import { Reminder } from "./store";
import { styles } from "./styles";


export function ListRemindersScreen() {
    const reminders = useSelector<any>(state => state.reminders.reminders) as Reminder[]

    return (
        <ReminderList reminders={reminders}></ReminderList>
    )
}

export function ReminderList({ reminders }: { reminders: Reminder[] }) {
    let content;
    if (reminders.length === 0) {
        content = <Text>No reminders yet, why not create one?</Text>
    }
    else {
        content = reminders.map((r) => <View key={r.id}><Text>
            Reminder at {r.time.toFormat("HH:mm")}
        </Text></View>)
    }

    return (
        <View style={styles.container}>
            {content}
        </View>
    )
}
