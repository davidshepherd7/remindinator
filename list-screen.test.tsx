import React from "react";
import { ReminderList } from "./list-screen";
import { mkReminder } from "./store";
import * as renderer from 'react-test-renderer';


describe(`ListReminders`, () => {
    it('renders', () => {
        const reminders = [mkReminder("10", "30"), mkReminder("12", "0")]
        const tree = renderer.create(<ReminderList reminders={reminders} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
