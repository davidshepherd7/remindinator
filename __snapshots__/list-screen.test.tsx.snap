// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListReminders renders 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <Text>
      Reminder at 
      10:30
    </Text>
  </View>
  <View>
    <Text>
      Reminder at 
      12:00
    </Text>
  </View>
</View>
`;
